### ISSUE TITLE
crash in Pdb generation

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#2749]
We’re seeing a (Watson... *don't tell*... tracked as DevDiv 1167556) crash in PDB generation in code that was added recently.

The method is at https://github.com/dotnet/roslyn/blob/2151da1a32a1d27e4d9cdc09d58172a141c1caa4/src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs#L625-L650

```cs
        private string GetAssemblyReferenceAlias(IAssemblyReference assembly, HashSet<string> declaredExternAliases)
        {
            var allAliases = _metadataWriter.Context.Module.GetAssemblyReferenceAliases(_metadataWriter.Context);
            foreach (AssemblyReferenceAlias alias in allAliases)
            {
                // Multiple aliases may be given to an assembly reference.
                // We find one that is in scope (was imported via extern alias directive).
                // If multiple are in scope then use the first one.

                // NOTE: Dev12 uses the one that appeared in source, whereas we use
                // the first one that COULD have appeared in source.  (DevDiv #913022)
                // The reason we're not just using the alias from the syntax is that
                // it is non-trivial to locate.  In particular, since "." may be used in
                // place of "::", determining whether the first identifier in the name is
                // the alias requires binding.  For example, "using A.B;" could refer to
                // either "A::B" or "global::A.B".

                if (assembly == alias.Assembly && declaredExternAliases.Contains(alias.Name))
                {
                    return alias.Name;
                }
            }

            // no alias defined in scope for given assembly -> error in compiler
            throw ExceptionUtilities.Unreachable;
        }
```

The callsites are at https://github.com/dotnet/roslyn/blob/2151da1a32a1d27e4d9cdc09d58172a141c1caa4/src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs#L521-L533

```cs
                if (import.AliasOpt != null)
                {
                    return (import.TargetAssemblyOpt != null) ?
                        "A" + import.AliasOpt + " E" + namespaceName + " " + GetAssemblyReferenceAlias(import.TargetAssemblyOpt, declaredExternAliasesOpt) :
                        "A" + import.AliasOpt + " U" + namespaceName;
                }
                else
                {
                    return (import.TargetAssemblyOpt != null) ?
                        "E" + namespaceName + " " + GetAssemblyReferenceAlias(import.TargetAssemblyOpt, declaredExternAliasesOpt) :
                        "U" + namespaceName;
                }
```

The callsites can pass `null` for `declaredExternAliases`, but the method fails if that is the case.

**Comment by gafter** [comment]
@tmat You added this code recently... any advice?

**Comment by tmat** [comment]
@gafter I'll take a look.

**Comment by tmat** [comment]
Seems like a duplicate of #1764

**Comment by gafter** [comment]
@tmat I think the code identified above is currently in the compiler.

**Comment by tmat** [comment]
@gafter Yes, however the code above is correct - it's just missing an assert that on this code path ```declaredExternAliasesOpt``` is expected to be non-null. I have added such assert. The root cause why it was null was fixed by #1764.
